source/opt/inline_pass.cpp

```diff
@@ -402,23 +468,117 @@ bool InlinePass::Inline(ir::Function* func) {
   return modified;
 }
 
-bool InlinePass::IsInlinableFunction(const ir::Function* func) {
-  // We can only inline a function if it has blocks.
-  if (func->cbegin() == func->cend())
+bool InlinePass::hasMultipleReturns(ir::Function* func) {
+  bool seenReturn = false;
+  bool multipleReturns = false;
+  for (auto& blk : *func) {
+    auto lii = blk.cend();
+    lii--;
+    if (lii->opcode() == SpvOpReturn || lii->opcode() == SpvOpReturnValue) {
+      if (seenReturn) {
+        multipleReturns = true;
+        break;
+      }
+      seenReturn = true;
+    }
+  }
+  return multipleReturns;
+}
+
+void InlinePass::ComputeStructuredSuccessors(ir::Function* func) {
+  // If header, make merge block first successor.
+  for (auto& blk : *func) {
+    auto mii = blk.end();
+    mii--;
+    uint32_t mbid = 0;
+    if (mii != blk.begin()) {
```

Comment:
> A vaild basic block always has at least two instructions: the label and the terminator instruction

That's what I originally thought too. But then reality in the form of a runtime abort stepped in.

I had forgotten that on the opt side, the OpLabel is kept separate from all the other instructions in a block. So when iterating over a block using begin/end, the OpLabel doesn't appear. Hence the extra logic.